src/Symfony/Component/HttpKernel/EventListener/CacheAttributeListener.php

```diff
@@ -0,0 +1,186 @@
+            return;
+        }
+        $response = $event->getResponse();
+
+        // http://tools.ietf.org/html/draft-ietf-httpbis-p4-conditional-12#section-3.1
+        if (!\in_array($response->getStatusCode(), [200, 203, 300, 301, 302, 304, 404, 410])) {
+            unset($this->lastModified[$request]);
+            unset($this->etags[$request]);
+
+            return;
+        }
+
+        if (isset($this->lastModified[$request]) && !$response->headers->has('Last-Modified')) {
+            $response->setLastModified($this->lastModified[$request]);
+        }
+
+        if (isset($this->etags[$request]) && !$response->headers->has('Etag')) {
+            $response->setEtag($this->etags[$request]);
+        }
+
+        unset($this->lastModified[$request]);
+        unset($this->etags[$request]);
+        $hasVary = $response->headers->has('Vary');
+
+        foreach (array_reverse($attributes) as $cache) {
+            if (null !== $cache->smaxage && !$response->headers->hasCacheControlDirective('s-maxage')) {
+                $response->setSharedMaxAge($this->toSeconds($cache->smaxage));
+            }
+
+            if ($cache->mustRevalidate) {
+                $response->headers->addCacheControlDirective('must-revalidate');
+            }
+
+            if (null !== $cache->maxage && !$response->headers->hasCacheControlDirective('max-age')) {
+                $response->setMaxAge($this->toSeconds($cache->maxage));
+            }
+
+            if (null !== $cache->maxStale && !$response->headers->hasCacheControlDirective('max-stale')) {
+                $response->headers->addCacheControlDirective('max-stale', $this->toSeconds($cache->maxStale));
+            }
+
+            if (null !== $cache->staleWhileRevalidate && !$response->headers->hasCacheControlDirective('stale-while-revalidate')) {
+                $response->headers->addCacheControlDirective('stale-while-revalidate', $this->toSeconds($cache->staleWhileRevalidate));
+            }
+
+            if (null !== $cache->staleIfError && !$response->headers->hasCacheControlDirective('stale-if-error')) {
+                $response->headers->addCacheControlDirective('stale-if-error', $this->toSeconds($cache->staleIfError));
+            }
+
+            if (null !== $cache->expires && !$response->headers->has('Expires')) {
+                $response->setExpires(new \DateTimeImmutable('@'.strtotime($cache->expires, time())));
+            }
+
+            if (!$hasVary && $cache->vary) {
+                $response->setVary($cache->vary, false);
+            }
+        }
+
+        foreach ($attributes as $cache) {
+            if (true === $cache->public) {
+                $response->setPublic();
+            }
+
+            if (false === $cache->public) {
+                $response->setPrivate();
+            }
+        }
+    }
+
+    public static function getSubscribedEvents(): array
+    {
+        return [
+            KernelEvents::CONTROLLER => 'onKernelController',
```

Comment:
perhaps out of topic, but will this work?

```suggestion
            ControllerEvent::class => 'onKernelController',
```

---

also 2 questions (regarding the request/response flow):

- what about defining a default priority here? (not `0`) to know precisely the order of execution
- following upper question; with all internals events subscribers/listeners, ca we assert the order of priority of each code?

Reply:
> what about defining a default priority here? (not `0`) to know precisely the order of execution

I think, this is a good idea. Zero would be the default for userland listeners and I suppose we want `onKernelController` to run before and `onKernelResponse` after them.

Reply:
I went with 10 and -10

Reply:
Using `ControllerEvent::class` should work, but I prefer keeping consistency with other listeners.